c10/core/ScalarType.h

```diff
@@ -193,19 +193,25 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
   if (isComplexType(a) || isComplexType(b)) {
     AT_ERROR("promoteTypes with complex numbers is not handled yet; figure out what the correct rules should be");
   }
+
+  // this matrix has to be consistent with AT_FORALL_SCALAR_TYPES_WITH_COMPLEX so that's why we have to add
+  // undefined as we are not sure what is the corrent values for the type promotions in complex type cases.
```

Comment:
I don't think you should call these 'ud' in the table -- that overload is confusing.  Just mark them as something like c2, c4, c8 or something and have them all map to ud.